src/dbnode/storage/flush.go

```diff
@@ -102,28 +105,69 @@ func (m *flushManager) Flush(
 
 	defer m.setState(flushManagerIdle)
 
-	// create flush-er
-	flushPersist, err := m.pm.StartFlushPersist()
+	namespaces, err := m.database.GetOwnedNamespaces()
 	if err != nil {
 		return err
 	}
 
-	namespaces, err := m.database.GetOwnedNamespaces()
+	// Perform three separate loops through all the namespaces so that we can
+	// emit better gauges, i.e. all the flushing for all the namespaces happens
+	// at once, then all the cold flushes, then all the snapshotting. This is
+	// also slightly better semantically because flushing should take priority
+	// over cold flushes and snapshotting.
+	//
+	// In addition, we need to make sure that for any given shard/blockStart
+	// combination, we attempt a flush and then a cold flush before a snapshot
+	// as the snapshotting process will attempt to snapshot any unflushed blocks
+	// which would be wasteful if the block is already flushable.
+	multiErr := xerrors.NewMultiError()
+	if err = m.dataWarmFlush(namespaces, tickStart, dbBootstrapStateAtTickStart); err != nil {
+		multiErr = multiErr.Add(err)
+	}
+
+	rotatedCommitlogID, err := m.commitlog.RotateLogs()
+	if err == nil {
+		if err = m.dataColdFlush(namespaces, tickStart); err != nil {
+			multiErr = multiErr.Add(err)
+			// If cold flush fails, we can't proceed to snapshotting because
+			// commit log logic just checks for a successful snapshot checkpoint
+			// file. Therefore if a cold flush fails and a snapshot succeeds,
+			// the writes from the failed cold flush might be lost when commit
+			// logs get cleaned up.
```

Comment:
"get cleaned up, leaving the node in an undurable state such that if it restarted it would not be able to recover the cold writes from its commit log"

I know I'm kind of dragging this stuff out, but better to be verbose here because eventually we'll forget

Reply:
Sure.